src/puppetlabs/dujour/version_check.clj

```diff
@@ -0,0 +1,127 @@
+
+(schema/defn get-coords :- ProductCoords
+  [product-name :- ProductName]
+  (condp schema-match? product-name
+    schema/Str {:group-id default-group-id
+                :artifact-id product-name}
+    ProductCoords product-name))
+
+(schema/defn version* :- schema/Str
+  "Get the version number of this installation."
+  [group-id artifact-id]
+  (version/get-version group-id artifact-id))
+
+(def version
+  "Get the version number of this installation."
+  (memoize version*))
+
+(schema/defn ^:always-validate update-info :- (schema/maybe UpdateInfo)
+  "Make a request to the puppetlabs server to determine the latest available
+  version. Returns the JSON object received from the server, which
+  is expected to be a map containing keys `:version`, `:newer`, and `:link`.
+  Returns `nil` if the request does not succeed for some reason."
+  [request-values :- RequestValues
+   update-server-url :- schema/Str]
+  (let [product-name (:product-name request-values)
+        {:keys [group-id artifact-id]} (get-coords product-name)
+        current-version (version group-id artifact-id)
+        version-data {:version current-version}
+        query-string (ring-codec/form-encode version-data)
+        url (format "%s?product=%s&group=%s&%s" update-server-url artifact-id group-id query-string)
+        {:keys [status body] :as resp} (client/get url
+                                                   {:headers {"Accept" "application/json"}
+                                                    :as :text})]
+    (if (= status 200)
+      (json/parse-string body true)
+      (sling/throw+ {:type ::update-request-failed
+                     :message resp}))))
+
+
+(defn validate-config!
+  [request-values update-server-url]
+  ;; if this ends up surfacing error messages that aren't very user-friendly,
+  ;; we can improve the validation logic.
+  (schema/validate RequestValues request-values)
+  (schema/validate (schema/maybe schema/Str) update-server-url))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Public
+
+(defn check-for-updates!
+  "This will fetch the latest version number and log if the system
+  is out of date."
+  [request-values update-server-url]
+  (log/debugf "Checking for newer versions of %s" (:product-name request-values))
+  (let [update-server-url             (or update-server-url default-update-server-url)
+        {:keys [version newer link]}  (try
+                                        (update-info request-values update-server-url)
+                                        (catch Throwable e
+                                          (log/debug e (format "Could not retrieve update information (%s)" update-server-url))))
+        link-str (if link
+                   (format " Visit %s for details." link)
+                   "")
+        update-msg (format "Newer version %s is available!%s" version link-str)]
+    (when newer
+      (log/info update-msg))))
+
+(defn version-check
```

Comment:
Out of curiosity, is there a reason to have two functions in the 'public API' here?  It seems like (concerns about test implementation and threading implementation aside), we'd want to be steering the user towards one or the other of these, rather than having them both live in 'public'.

FWIW I'd vote for the public function to be called `check-for-updates!`, and we should probably add something to the docstring that indicates that it will happen asynchronously and indicate that the function returns `nil` or something.